### PR TITLE
Arnej/configurable distance metrics

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@geirst please review
@havardpe FYI

### CODE REVIEW

**Review by geirst — COMMENTED** [review]
Overall looks good. We should add a new unit tests (gtest) for the distance functions (euclidean is already tested in hw accelerated code).

File: searchlib/src/vespa/searchlib/tensor/distance_functions.h [hunk 1]
```diff
@@ -25,10 +26,111 @@ class SquaredEuclideanDistance : public DistanceFunction {
...
+
+    const vespalib::hwaccelrated::IAccelrated & _computer;
+};
+
+template class AngularDistance<float>;
+template class AngularDistance<double>;
+
+/**
+ * Calculates great-circle distance between Latitude/Longitude pairs,
+ * measured in degrees
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply searchlib/src/vespa/searchlib/tensor/distance_functions.h]: pointer to wikipedia added

File: config-model/src/test/java/com/yahoo/searchdefinition/document/HnswIndexParamsTestCase.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply config-model/src/test/java/com/yahoo/searchdefinition/document/HnswIndexParamsTestCase.java]: copyright added

File: searchlib/src/vespa/searchlib/tensor/distance_function_factory.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+#pragma once
+
+#include "distance_function.h"
+#include <vespa/eval/eval/value_type.h>
+#include <vespa/searchcommon/attribute/distance_metric.h>
+
+namespace search::tensor {
+
+DistanceFunction::UP
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply searchlib/src/vespa/searchlib/tensor/distance_function_factory.h]: function description added.

File: searchlib/src/vespa/searchlib/tensor/nearest_neighbor_index.h [hunk 4]
```diff
@@ -35,6 +36,7 @@ class NearestNeighborIndex {
                                              vespalib::tensor::TypedCells vector,
                                              uint32_t explore_k) const = 0;
 
+    virtual DistanceFunction *distance_function() const = 0;
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply searchlib/src/vespa/searchlib/tensor/nearest_neighbor_index.h]: constified.

File: searchlib/src/vespa/searchlib/attribute/configconverter.cpp [hunk 5]
```diff
@@ -74,8 +74,22 @@ ConfigConverter::convert(const AttributesConfig::Attribute & cfg)
...
+            dm = DistanceMetric::Angular;
+            break;
+        case CfgDm::GEODEGREES:
+            dm = DistanceMetric::GeoDegrees;
+            break;
+        }
         retval.set_hnsw_index_params(HnswIndexParams(cfg.index.hnsw.maxlinkspernode,
-                                                     cfg.index.hnsw.neighborstoexploreatinsert));
+                                                     cfg.index.hnsw.neighborstoexploreatinsert,
+                                                     dm));
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply searchlib/src/vespa/searchlib/attribute/configconverter.cpp]: unit test extended

File: searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_blueprint.cpp [hunk 6]
```diff
@@ -6,10 +6,47 @@
...
+namespace {
+
+template<typename LCT, typename RCT>
+void
+convert_cells(std::unique_ptr<DenseTensorView> &original, vespalib::eval::ValueType want_type)
+{
+    auto old_cells = original->cellsRef().typify<LCT>();
+    std::vector<RCT> new_cells;
+    new_cells.reserve(old_cells.size());
+    for (LCT value : old_cells) {
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_blueprint.cpp]: Here I feel using the actual type makes it clearer what is going on, and using "auto" would not save any typing either.

**Comment by arnej27959** [comment]
added a unit test that also checks geo distances around the globe.
@geirst PTAL

File: searchlib/src/tests/tensor/distance_functions/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Copyright 2020 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline searchlib/src/tests/tensor/distance_functions/CMakeLists.txt]: Please update copyright.

File: searchlib/src/tests/tensor/distance_functions/distance_functions_test.cpp [hunk 8]
```diff
@@ -0,0 +1,196 @@
+// Copyright 2020 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline searchlib/src/tests/tensor/distance_functions/distance_functions_test.cpp]: Please update copyright.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**